{
  "cells": [
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "bW8EhV-RJLos"
      },
      "outputs": [],
      "source": [
        "import pandas as pd\n",
        "import re\n",
        "\n",
        "def get_vcf(vcf_path):\n",
        "    with open(vcf_path, \"rt\") as ifile:\n",
        "          for line in ifile:\n",
        "            if line.startswith(\"#CHROM\"):\n",
        "                  vcf_names = [x for x in line.split('\\t')]\n",
        "                  break\n",
        "    ifile.close()\n",
        "    return vcf_names\n",
        "\n",
        "\n",
        "def vcf_edit(vcf_name):\n",
        "  names = get_vcf(vcf_name)\n",
        "  vcfDF = pd.read_csv(vcf_name, comment='#', delim_whitespace=True,header=None, names=names)\n",
        "  vcfRest=vcfDF.loc[:,[\"#CHROM\",\"POS\",\"ID\",\"REF\",\"ALT\"]]\n",
        "  info = vcfDF['INFO\\n'].str.split(';', expand=True)\n",
        "\n",
        "  for row in info.itertuples():\n",
        "    if 'ANN=' in row:\n",
        "      ann = info[11]\n",
        "    else: \n",
        "      print('ww')\n",
        "      ann = info[12]\n",
        "    \n",
        "\n",
        "    copyOfann = ann.copy()\n",
        "    for i in ann:\n",
        "      if i.startswith('ANN'):\n",
        "        ann_ext = ann.str.replace(',','\\n',regex=True)\n",
        "        ann_ext = ann_ext.str.split(\"|\", expand=True)\n",
        "        ann_ext = ann_ext.iloc[:,[0,3,4,6,9,10]]\n",
        "        ann_ext.columns=[\"Nucleotide\",\"Gene\",\"GeneID\",\"TranscriptID\",\"Change\",\"Aminoacid\"]\n",
        "        result = ann_ext[\"Aminoacid\"].str.split('([A-Za-z]+)(\\d+)([A-Za-z]+)', expand=True)\n",
        "        result = result.loc[:,[1,3,2]]\n",
        "        result.columns=[\"ReferenceAA\",\"AlternativeAA\",\"Codon\"]\n",
        "        Frames = [vcfRest, ann_ext, result]\n",
        "        mergeDf = pd.concat([vcfRest, ann_ext, result], axis=1)\n",
        "        mergeDf.head()\n",
        "      \n",
        "        return mergeDf\n",
        "      \n",
        "    "
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 505
        },
        "id": "ENaH_bgCJNwO",
        "outputId": "9550048f-468d-4deb-dda3-924b3e439a8a"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "ww\n"
          ]
        },
        {
          "data": {
            "text/html": [
              "\n",
              "  <div id=\"df-54b5ba15-e4b2-41ec-8c5e-8452fe2d1384\">\n",
              "    <div class=\"colab-df-container\">\n",
              "      <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>#CHROM</th>\n",
              "      <th>POS</th>\n",
              "      <th>ID</th>\n",
              "      <th>REF</th>\n",
              "      <th>ALT</th>\n",
              "      <th>Nucleotide</th>\n",
              "      <th>Gene</th>\n",
              "      <th>GeneID</th>\n",
              "      <th>TranscriptID</th>\n",
              "      <th>Change</th>\n",
              "      <th>Aminoacid</th>\n",
              "      <th>ReferenceAA</th>\n",
              "      <th>AlternativeAA</th>\n",
              "      <th>Codon</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>1</td>\n",
              "      <td>925952</td>\n",
              "      <td>1019397</td>\n",
              "      <td>G</td>\n",
              "      <td>A</td>\n",
              "      <td>ANN=A</td>\n",
              "      <td>SAMD11</td>\n",
              "      <td>ENSG00000187634</td>\n",
              "      <td>ENST00000618323.5</td>\n",
              "      <td>c.548G&gt;A</td>\n",
              "      <td>p.Gly183Glu</td>\n",
              "      <td>Gly</td>\n",
              "      <td>Glu</td>\n",
              "      <td>183</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>1</td>\n",
              "      <td>925956</td>\n",
              "      <td>1543320</td>\n",
              "      <td>C</td>\n",
              "      <td>T</td>\n",
              "      <td>None</td>\n",
              "      <td>None</td>\n",
              "      <td>None</td>\n",
              "      <td>None</td>\n",
              "      <td>None</td>\n",
              "      <td>None</td>\n",
              "      <td>None</td>\n",
              "      <td>None</td>\n",
              "      <td>None</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>1</td>\n",
              "      <td>925969</td>\n",
              "      <td>1648427</td>\n",
              "      <td>C</td>\n",
              "      <td>T</td>\n",
              "      <td>None</td>\n",
              "      <td>None</td>\n",
              "      <td>None</td>\n",
              "      <td>None</td>\n",
              "      <td>None</td>\n",
              "      <td>None</td>\n",
              "      <td>None</td>\n",
              "      <td>None</td>\n",
              "      <td>None</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>1</td>\n",
              "      <td>925976</td>\n",
              "      <td>1362713</td>\n",
              "      <td>T</td>\n",
              "      <td>C</td>\n",
              "      <td>None</td>\n",
              "      <td>None</td>\n",
              "      <td>None</td>\n",
              "      <td>None</td>\n",
              "      <td>None</td>\n",
              "      <td>None</td>\n",
              "      <td>None</td>\n",
              "      <td>None</td>\n",
              "      <td>None</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>1</td>\n",
              "      <td>925986</td>\n",
              "      <td>1568423</td>\n",
              "      <td>C</td>\n",
              "      <td>T</td>\n",
              "      <td>None</td>\n",
              "      <td>None</td>\n",
              "      <td>None</td>\n",
              "      <td>None</td>\n",
              "      <td>None</td>\n",
              "      <td>None</td>\n",
              "      <td>None</td>\n",
              "      <td>None</td>\n",
              "      <td>None</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>462</th>\n",
              "      <td>1</td>\n",
              "      <td>943804</td>\n",
              "      <td>955433</td>\n",
              "      <td>C</td>\n",
              "      <td>T</td>\n",
              "      <td>ORIGIN=1</td>\n",
              "      <td>None</td>\n",
              "      <td>None</td>\n",
              "      <td>None</td>\n",
              "      <td>None</td>\n",
              "      <td>None</td>\n",
              "      <td>None</td>\n",
              "      <td>None</td>\n",
              "      <td>None</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>463</th>\n",
              "      <td>1</td>\n",
              "      <td>943805</td>\n",
              "      <td>1611881</td>\n",
              "      <td>C</td>\n",
              "      <td>G</td>\n",
              "      <td>None</td>\n",
              "      <td>None</td>\n",
              "      <td>None</td>\n",
              "      <td>None</td>\n",
              "      <td>None</td>\n",
              "      <td>None</td>\n",
              "      <td>None</td>\n",
              "      <td>None</td>\n",
              "      <td>None</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>464</th>\n",
              "      <td>1</td>\n",
              "      <td>943813</td>\n",
              "      <td>1476856</td>\n",
              "      <td>G</td>\n",
              "      <td>A</td>\n",
              "      <td>None</td>\n",
              "      <td>None</td>\n",
              "      <td>None</td>\n",
              "      <td>None</td>\n",
              "      <td>None</td>\n",
              "      <td>None</td>\n",
              "      <td>None</td>\n",
              "      <td>None</td>\n",
              "      <td>None</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>465</th>\n",
              "      <td>1</td>\n",
              "      <td>943815</td>\n",
              "      <td>1160452</td>\n",
              "      <td>C</td>\n",
              "      <td>T</td>\n",
              "      <td>None</td>\n",
              "      <td>None</td>\n",
              "      <td>None</td>\n",
              "      <td>None</td>\n",
              "      <td>None</td>\n",
              "      <td>None</td>\n",
              "      <td>None</td>\n",
              "      <td>None</td>\n",
              "      <td>None</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>466</th>\n",
              "      <td>1</td>\n",
              "      <td>943823</td>\n",
              "      <td>1553569</td>\n",
              "      <td>A</td>\n",
              "      <td>C</td>\n",
              "      <td>None</td>\n",
              "      <td>None</td>\n",
              "      <td>None</td>\n",
              "      <td>None</td>\n",
              "      <td>None</td>\n",
              "      <td>None</td>\n",
              "      <td>None</td>\n",
              "      <td>None</td>\n",
              "      <td>None</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>467 rows × 14 columns</p>\n",
              "</div>\n",
              "      <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-54b5ba15-e4b2-41ec-8c5e-8452fe2d1384')\"\n",
              "              title=\"Convert this dataframe to an interactive table.\"\n",
              "              style=\"display:none;\">\n",
              "        \n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "       width=\"24px\">\n",
              "    <path d=\"M0 0h24v24H0V0z\" fill=\"none\"/>\n",
              "    <path d=\"M18.56 5.44l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94zm-11 1L8.5 8.5l.94-2.06 2.06-.94-2.06-.94L8.5 2.5l-.94 2.06-2.06.94zm10 10l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94z\"/><path d=\"M17.41 7.96l-1.37-1.37c-.4-.4-.92-.59-1.43-.59-.52 0-1.04.2-1.43.59L10.3 9.45l-7.72 7.72c-.78.78-.78 2.05 0 2.83L4 21.41c.39.39.9.59 1.41.59.51 0 1.02-.2 1.41-.59l7.78-7.78 2.81-2.81c.8-.78.8-2.07 0-2.86zM5.41 20L4 18.59l7.72-7.72 1.47 1.35L5.41 20z\"/>\n",
              "  </svg>\n",
              "      </button>\n",
              "      \n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      flex-wrap:wrap;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "      <script>\n",
              "        const buttonEl =\n",
              "          document.querySelector('#df-54b5ba15-e4b2-41ec-8c5e-8452fe2d1384 button.colab-df-convert');\n",
              "        buttonEl.style.display =\n",
              "          google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "        async function convertToInteractive(key) {\n",
              "          const element = document.querySelector('#df-54b5ba15-e4b2-41ec-8c5e-8452fe2d1384');\n",
              "          const dataTable =\n",
              "            await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                     [key], {});\n",
              "          if (!dataTable) return;\n",
              "\n",
              "          const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "            '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "            + ' to learn more about interactive tables.';\n",
              "          element.innerHTML = '';\n",
              "          dataTable['output_type'] = 'display_data';\n",
              "          await google.colab.output.renderOutput(dataTable, element);\n",
              "          const docLink = document.createElement('div');\n",
              "          docLink.innerHTML = docLinkHtml;\n",
              "          element.appendChild(docLink);\n",
              "        }\n",
              "      </script>\n",
              "    </div>\n",
              "  </div>\n",
              "  "
            ],
            "text/plain": [
              "     #CHROM     POS       ID REF ALT Nucleotide    Gene           GeneID  \\\n",
              "0         1  925952  1019397   G   A      ANN=A  SAMD11  ENSG00000187634   \n",
              "1         1  925956  1543320   C   T       None    None             None   \n",
              "2         1  925969  1648427   C   T       None    None             None   \n",
              "3         1  925976  1362713   T   C       None    None             None   \n",
              "4         1  925986  1568423   C   T       None    None             None   \n",
              "..      ...     ...      ...  ..  ..        ...     ...              ...   \n",
              "462       1  943804   955433   C   T   ORIGIN=1    None             None   \n",
              "463       1  943805  1611881   C   G       None    None             None   \n",
              "464       1  943813  1476856   G   A       None    None             None   \n",
              "465       1  943815  1160452   C   T       None    None             None   \n",
              "466       1  943823  1553569   A   C       None    None             None   \n",
              "\n",
              "          TranscriptID    Change    Aminoacid ReferenceAA AlternativeAA Codon  \n",
              "0    ENST00000618323.5  c.548G>A  p.Gly183Glu         Gly           Glu   183  \n",
              "1                 None      None         None        None          None  None  \n",
              "2                 None      None         None        None          None  None  \n",
              "3                 None      None         None        None          None  None  \n",
              "4                 None      None         None        None          None  None  \n",
              "..                 ...       ...          ...         ...           ...   ...  \n",
              "462               None      None         None        None          None  None  \n",
              "463               None      None         None        None          None  None  \n",
              "464               None      None         None        None          None  None  \n",
              "465               None      None         None        None          None  None  \n",
              "466               None      None         None        None          None  None  \n",
              "\n",
              "[467 rows x 14 columns]"
            ]
          },
          "execution_count": 2,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "data = vcf_edit('/content/orj500_new4.vcf')\n",
        "data"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "KQPUMgt3JRNX"
      },
      "outputs": [],
      "source": [
        "import csv\n",
        " \n",
        "lines =[]\n",
        "with open(\"orj500_new4.vcf\") as file:\n",
        "    tsv_file = csv.reader(file, delimiter=\"\\t\")\n",
        "    for line in tsv_file:\n",
        "      if line[0][0] != '#':\n",
        "        lines.append(line)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "bnOcOyY3JNth"
      },
      "outputs": [],
      "source": [
        "updated_lines=[]\n",
        "for line in lines: \n",
        "  indv_line=[]\n",
        "  indv_line.append(line[0])\n",
        "  indv_line.append(line[1])\n",
        "  indv_line.append(line[2])\n",
        "  indv_line.append(line[3])\n",
        "  indv_line.append(line[4])\n",
        "  seperated = line[7].split(';')\n",
        "  for element in seperated: \n",
        "    if element.find('ANN') != -1: \n",
        "      annotation= element\n",
        "  annotation_sep = annotation.split('|')\n",
        "  indv_line.append(annotation_sep[1])\n",
        "  indv_line.append(annotation_sep[2])\n",
        "  indv_line.append(annotation_sep[3])\n",
        "  indv_line.append(annotation_sep[4])\n",
        "  indv_line.append(annotation_sep[6])\n",
        "  indv_line.append(annotation_sep[7])\n",
        "  indv_line.append(annotation_sep[9])\n",
        "  indv_line.append(annotation_sep[10])\n",
        "  updated_lines.append(indv_line)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "W2UAjjU0JNrB"
      },
      "outputs": [],
      "source": [
        "df = pd.DataFrame(updated_lines)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 487
        },
        "id": "Fw0vjnxtJNou",
        "outputId": "5e6066df-f20d-442b-8458-4acc444f8c6b"
      },
      "outputs": [
        {
          "data": {
            "text/html": [
              "\n",
              "  <div id=\"df-4e736b66-8659-4e30-9026-055adb41f752\">\n",
              "    <div class=\"colab-df-container\">\n",
              "      <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>CHROM</th>\n",
              "      <th>POS</th>\n",
              "      <th>ID</th>\n",
              "      <th>REF</th>\n",
              "      <th>ALT</th>\n",
              "      <th>Variant_type</th>\n",
              "      <th>QUAL</th>\n",
              "      <th>GENE</th>\n",
              "      <th>GeneID</th>\n",
              "      <th>TranscriptID</th>\n",
              "      <th>CODING</th>\n",
              "      <th>Change</th>\n",
              "      <th>Aminoacid</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>1</td>\n",
              "      <td>925952</td>\n",
              "      <td>1019397</td>\n",
              "      <td>G</td>\n",
              "      <td>A</td>\n",
              "      <td>missense_variant</td>\n",
              "      <td>MODERATE</td>\n",
              "      <td>SAMD11</td>\n",
              "      <td>ENSG00000187634</td>\n",
              "      <td>ENST00000618323.5</td>\n",
              "      <td>protein_coding</td>\n",
              "      <td>c.548G&gt;A</td>\n",
              "      <td>p.Gly183Glu</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>1</td>\n",
              "      <td>925956</td>\n",
              "      <td>1543320</td>\n",
              "      <td>C</td>\n",
              "      <td>T</td>\n",
              "      <td>synonymous_variant</td>\n",
              "      <td>LOW</td>\n",
              "      <td>SAMD11</td>\n",
              "      <td>ENSG00000187634</td>\n",
              "      <td>ENST00000618323.5</td>\n",
              "      <td>protein_coding</td>\n",
              "      <td>c.552C&gt;T</td>\n",
              "      <td>p.Ile184Ile</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>1</td>\n",
              "      <td>925969</td>\n",
              "      <td>1648427</td>\n",
              "      <td>C</td>\n",
              "      <td>T</td>\n",
              "      <td>missense_variant</td>\n",
              "      <td>MODERATE</td>\n",
              "      <td>SAMD11</td>\n",
              "      <td>ENSG00000187634</td>\n",
              "      <td>ENST00000618323.5</td>\n",
              "      <td>protein_coding</td>\n",
              "      <td>c.565C&gt;T</td>\n",
              "      <td>p.Pro189Ser</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>1</td>\n",
              "      <td>925976</td>\n",
              "      <td>1362713</td>\n",
              "      <td>T</td>\n",
              "      <td>C</td>\n",
              "      <td>missense_variant</td>\n",
              "      <td>MODERATE</td>\n",
              "      <td>SAMD11</td>\n",
              "      <td>ENSG00000187634</td>\n",
              "      <td>ENST00000618323.5</td>\n",
              "      <td>protein_coding</td>\n",
              "      <td>c.572T&gt;C</td>\n",
              "      <td>p.Ile191Thr</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>1</td>\n",
              "      <td>925986</td>\n",
              "      <td>1568423</td>\n",
              "      <td>C</td>\n",
              "      <td>T</td>\n",
              "      <td>synonymous_variant</td>\n",
              "      <td>LOW</td>\n",
              "      <td>SAMD11</td>\n",
              "      <td>ENSG00000187634</td>\n",
              "      <td>ENST00000618323.5</td>\n",
              "      <td>protein_coding</td>\n",
              "      <td>c.582C&gt;T</td>\n",
              "      <td>p.Cys194Cys</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>462</th>\n",
              "      <td>1</td>\n",
              "      <td>943804</td>\n",
              "      <td>955433</td>\n",
              "      <td>C</td>\n",
              "      <td>T</td>\n",
              "      <td>missense_variant</td>\n",
              "      <td>MODERATE</td>\n",
              "      <td>SAMD11</td>\n",
              "      <td>ENSG00000187634</td>\n",
              "      <td>ENST00000455979.1</td>\n",
              "      <td>protein_coding</td>\n",
              "      <td>c.1274C&gt;T</td>\n",
              "      <td>p.Ala425Val</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>463</th>\n",
              "      <td>1</td>\n",
              "      <td>943805</td>\n",
              "      <td>1611881</td>\n",
              "      <td>C</td>\n",
              "      <td>G</td>\n",
              "      <td>synonymous_variant</td>\n",
              "      <td>LOW</td>\n",
              "      <td>SAMD11</td>\n",
              "      <td>ENSG00000187634</td>\n",
              "      <td>ENST00000455979.1</td>\n",
              "      <td>protein_coding</td>\n",
              "      <td>c.1275C&gt;G</td>\n",
              "      <td>p.Ala425Ala</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>464</th>\n",
              "      <td>1</td>\n",
              "      <td>943813</td>\n",
              "      <td>1476856</td>\n",
              "      <td>G</td>\n",
              "      <td>A</td>\n",
              "      <td>missense_variant</td>\n",
              "      <td>MODERATE</td>\n",
              "      <td>SAMD11</td>\n",
              "      <td>ENSG00000187634</td>\n",
              "      <td>ENST00000455979.1</td>\n",
              "      <td>protein_coding</td>\n",
              "      <td>c.1283G&gt;A</td>\n",
              "      <td>p.Arg428Lys</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>465</th>\n",
              "      <td>1</td>\n",
              "      <td>943815</td>\n",
              "      <td>1160452</td>\n",
              "      <td>C</td>\n",
              "      <td>T</td>\n",
              "      <td>missense_variant</td>\n",
              "      <td>MODERATE</td>\n",
              "      <td>SAMD11</td>\n",
              "      <td>ENSG00000187634</td>\n",
              "      <td>ENST00000455979.1</td>\n",
              "      <td>protein_coding</td>\n",
              "      <td>c.1285C&gt;T</td>\n",
              "      <td>p.Arg429Cys</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>466</th>\n",
              "      <td>1</td>\n",
              "      <td>943823</td>\n",
              "      <td>1553569</td>\n",
              "      <td>A</td>\n",
              "      <td>C</td>\n",
              "      <td>synonymous_variant</td>\n",
              "      <td>LOW</td>\n",
              "      <td>SAMD11</td>\n",
              "      <td>ENSG00000187634</td>\n",
              "      <td>ENST00000455979.1</td>\n",
              "      <td>protein_coding</td>\n",
              "      <td>c.1293A&gt;C</td>\n",
              "      <td>p.Gly431Gly</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>467 rows × 13 columns</p>\n",
              "</div>\n",
              "      <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-4e736b66-8659-4e30-9026-055adb41f752')\"\n",
              "              title=\"Convert this dataframe to an interactive table.\"\n",
              "              style=\"display:none;\">\n",
              "        \n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "       width=\"24px\">\n",
              "    <path d=\"M0 0h24v24H0V0z\" fill=\"none\"/>\n",
              "    <path d=\"M18.56 5.44l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94zm-11 1L8.5 8.5l.94-2.06 2.06-.94-2.06-.94L8.5 2.5l-.94 2.06-2.06.94zm10 10l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94z\"/><path d=\"M17.41 7.96l-1.37-1.37c-.4-.4-.92-.59-1.43-.59-.52 0-1.04.2-1.43.59L10.3 9.45l-7.72 7.72c-.78.78-.78 2.05 0 2.83L4 21.41c.39.39.9.59 1.41.59.51 0 1.02-.2 1.41-.59l7.78-7.78 2.81-2.81c.8-.78.8-2.07 0-2.86zM5.41 20L4 18.59l7.72-7.72 1.47 1.35L5.41 20z\"/>\n",
              "  </svg>\n",
              "      </button>\n",
              "      \n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      flex-wrap:wrap;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "      <script>\n",
              "        const buttonEl =\n",
              "          document.querySelector('#df-4e736b66-8659-4e30-9026-055adb41f752 button.colab-df-convert');\n",
              "        buttonEl.style.display =\n",
              "          google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "        async function convertToInteractive(key) {\n",
              "          const element = document.querySelector('#df-4e736b66-8659-4e30-9026-055adb41f752');\n",
              "          const dataTable =\n",
              "            await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                     [key], {});\n",
              "          if (!dataTable) return;\n",
              "\n",
              "          const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "            '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "            + ' to learn more about interactive tables.';\n",
              "          element.innerHTML = '';\n",
              "          dataTable['output_type'] = 'display_data';\n",
              "          await google.colab.output.renderOutput(dataTable, element);\n",
              "          const docLink = document.createElement('div');\n",
              "          docLink.innerHTML = docLinkHtml;\n",
              "          element.appendChild(docLink);\n",
              "        }\n",
              "      </script>\n",
              "    </div>\n",
              "  </div>\n",
              "  "
            ],
            "text/plain": [
              "    CHROM     POS       ID REF ALT        Variant_type      QUAL    GENE  \\\n",
              "0       1  925952  1019397   G   A    missense_variant  MODERATE  SAMD11   \n",
              "1       1  925956  1543320   C   T  synonymous_variant       LOW  SAMD11   \n",
              "2       1  925969  1648427   C   T    missense_variant  MODERATE  SAMD11   \n",
              "3       1  925976  1362713   T   C    missense_variant  MODERATE  SAMD11   \n",
              "4       1  925986  1568423   C   T  synonymous_variant       LOW  SAMD11   \n",
              "..    ...     ...      ...  ..  ..                 ...       ...     ...   \n",
              "462     1  943804   955433   C   T    missense_variant  MODERATE  SAMD11   \n",
              "463     1  943805  1611881   C   G  synonymous_variant       LOW  SAMD11   \n",
              "464     1  943813  1476856   G   A    missense_variant  MODERATE  SAMD11   \n",
              "465     1  943815  1160452   C   T    missense_variant  MODERATE  SAMD11   \n",
              "466     1  943823  1553569   A   C  synonymous_variant       LOW  SAMD11   \n",
              "\n",
              "              GeneID       TranscriptID          CODING     Change  \\\n",
              "0    ENSG00000187634  ENST00000618323.5  protein_coding   c.548G>A   \n",
              "1    ENSG00000187634  ENST00000618323.5  protein_coding   c.552C>T   \n",
              "2    ENSG00000187634  ENST00000618323.5  protein_coding   c.565C>T   \n",
              "3    ENSG00000187634  ENST00000618323.5  protein_coding   c.572T>C   \n",
              "4    ENSG00000187634  ENST00000618323.5  protein_coding   c.582C>T   \n",
              "..               ...                ...             ...        ...   \n",
              "462  ENSG00000187634  ENST00000455979.1  protein_coding  c.1274C>T   \n",
              "463  ENSG00000187634  ENST00000455979.1  protein_coding  c.1275C>G   \n",
              "464  ENSG00000187634  ENST00000455979.1  protein_coding  c.1283G>A   \n",
              "465  ENSG00000187634  ENST00000455979.1  protein_coding  c.1285C>T   \n",
              "466  ENSG00000187634  ENST00000455979.1  protein_coding  c.1293A>C   \n",
              "\n",
              "       Aminoacid  \n",
              "0    p.Gly183Glu  \n",
              "1    p.Ile184Ile  \n",
              "2    p.Pro189Ser  \n",
              "3    p.Ile191Thr  \n",
              "4    p.Cys194Cys  \n",
              "..           ...  \n",
              "462  p.Ala425Val  \n",
              "463  p.Ala425Ala  \n",
              "464  p.Arg428Lys  \n",
              "465  p.Arg429Cys  \n",
              "466  p.Gly431Gly  \n",
              "\n",
              "[467 rows x 13 columns]"
            ]
          },
          "execution_count": 6,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "df.columns=[\"CHROM\",\"POS\",\"ID\",\"REF\",\"ALT\",\"Variant_type\",\"QUAL\",\"GENE\", \"GeneID\",\"TranscriptID\", \"CODING\",\"Change\",\"Aminoacid\"]\n",
        "df"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "jylGL7yRJNl7",
        "outputId": "f0511cb3-80ba-43e9-c45c-8351229e080f"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/\n",
            "Collecting pyranges\n",
            "  Downloading pyranges-0.0.120.tar.gz (687 kB)\n",
            "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m687.9/687.9 KB\u001b[0m \u001b[31m35.9 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25h  Preparing metadata (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "Requirement already satisfied: cython in /usr/local/lib/python3.9/dist-packages (from pyranges) (0.29.33)\n",
            "Requirement already satisfied: pandas in /usr/local/lib/python3.9/dist-packages (from pyranges) (1.3.5)\n",
            "Collecting ncls>=0.0.63\n",
            "  Downloading ncls-0.0.66.tar.gz (494 kB)\n",
            "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m494.7/494.7 KB\u001b[0m \u001b[31m44.3 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25h  Preparing metadata (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "Requirement already satisfied: tabulate in /usr/local/lib/python3.9/dist-packages (from pyranges) (0.8.10)\n",
            "Collecting sorted_nearest>=0.0.33\n",
            "  Downloading sorted_nearest-0.0.38.tar.gz (1.5 MB)\n",
            "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m1.5/1.5 MB\u001b[0m \u001b[31m30.7 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25h  Installing build dependencies ... \u001b[?25l\u001b[?25hdone\n",
            "  Getting requirements to build wheel ... \u001b[?25l\u001b[?25hdone\n",
            "  Preparing metadata (pyproject.toml) ... \u001b[?25l\u001b[?25hdone\n",
            "Collecting pyrle\n",
            "  Downloading pyrle-0.0.35.tar.gz (425 kB)\n",
            "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m425.3/425.3 KB\u001b[0m \u001b[31m39.1 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25h  Installing build dependencies ... \u001b[?25l\u001b[?25hdone\n",
            "  Getting requirements to build wheel ... \u001b[?25l\u001b[?25hdone\n",
            "  Preparing metadata (pyproject.toml) ... \u001b[?25l\u001b[?25hdone\n",
            "Requirement already satisfied: natsort in /usr/local/lib/python3.9/dist-packages (from pyranges) (5.5.0)\n",
            "Requirement already satisfied: numpy in /usr/local/lib/python3.9/dist-packages (from ncls>=0.0.63->pyranges) (1.22.4)\n",
            "Requirement already satisfied: python-dateutil>=2.7.3 in /usr/local/lib/python3.9/dist-packages (from pandas->pyranges) (2.8.2)\n",
            "Requirement already satisfied: pytz>=2017.3 in /usr/local/lib/python3.9/dist-packages (from pandas->pyranges) (2022.7.1)\n",
            "Requirement already satisfied: six>=1.5 in /usr/local/lib/python3.9/dist-packages (from python-dateutil>=2.7.3->pandas->pyranges) (1.15.0)\n",
            "Building wheels for collected packages: pyranges, ncls, sorted_nearest, pyrle\n",
            "  Building wheel for pyranges (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "  Created wheel for pyranges: filename=pyranges-0.0.120-py3-none-any.whl size=697163 sha256=6b8093f4c63e2ce35e06a77d7fa03cbdae1515af291507970ec4e3eef9b601bb\n",
            "  Stored in directory: /root/.cache/pip/wheels/6a/9a/6a/03a591e356aa16c6b447873e9d08dad1d66d7f579eaf99f1a3\n",
            "  Building wheel for ncls (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "  Created wheel for ncls: filename=ncls-0.0.66-cp39-cp39-linux_x86_64.whl size=2385459 sha256=05dc20c88125ef6c297e3b7c119d93b0571ce7d957b08ec289672ab2e973efff\n",
            "  Stored in directory: /root/.cache/pip/wheels/da/12/ee/345b9dc8224851d613fd67e9183a5d2742078d27480467bca2\n",
            "  Building wheel for sorted_nearest (pyproject.toml) ... \u001b[?25l\u001b[?25hdone\n",
            "  Created wheel for sorted_nearest: filename=sorted_nearest-0.0.38-cp39-cp39-linux_x86_64.whl size=5933220 sha256=b12d2d2bb6fff72e2a1a5968ad08126fe20ebeb62f9cbec39d54172fceccb38e\n",
            "  Stored in directory: /root/.cache/pip/wheels/eb/cb/84/a96707e1b437c7dc463cb2535e98f91d078644b5106fbedd60\n",
            "  Building wheel for pyrle (pyproject.toml) ... \u001b[?25l\u001b[?25hdone\n",
            "  Created wheel for pyrle: filename=pyrle-0.0.35-cp39-cp39-linux_x86_64.whl size=1715234 sha256=bf2c049c8f5dbaf9fdb73bab7496b25cc695f0b27ca4ee587ca58781986a4ee9\n",
            "  Stored in directory: /root/.cache/pip/wheels/f0/38/2c/be598f7dfbb6f344ef1e53d4302f570c8e37d3a039358285fd\n",
            "Successfully built pyranges ncls sorted_nearest pyrle\n",
            "Installing collected packages: sorted_nearest, ncls, pyrle, pyranges\n",
            "Successfully installed ncls-0.0.66 pyranges-0.0.120 pyrle-0.0.35 sorted_nearest-0.0.38\n"
          ]
        }
      ],
      "source": [
        "!pip install pyranges"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 423
        },
        "id": "XxDF199FJNZc",
        "outputId": "68abae3d-d3bf-4550-f365-1ebf33b7db77"
      },
      "outputs": [
        {
          "data": {
            "text/html": [
              "\n",
              "  <div id=\"df-d8f2e8d1-0936-4b98-a844-888fdebb7ef4\">\n",
              "    <div class=\"colab-df-container\">\n",
              "      <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Chromosome</th>\n",
              "      <th>Start</th>\n",
              "      <th>End</th>\n",
              "      <th>gene_id</th>\n",
              "      <th>transcript_id</th>\n",
              "      <th>protein_id</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>1</td>\n",
              "      <td>1471884</td>\n",
              "      <td>1472089</td>\n",
              "      <td>ENSG00000160072</td>\n",
              "      <td>ENST00000673477</td>\n",
              "      <td>ENSP00000500094</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>6</th>\n",
              "      <td>1</td>\n",
              "      <td>1477273</td>\n",
              "      <td>1477350</td>\n",
              "      <td>ENSG00000160072</td>\n",
              "      <td>ENST00000673477</td>\n",
              "      <td>ENSP00000500094</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>8</th>\n",
              "      <td>1</td>\n",
              "      <td>1478643</td>\n",
              "      <td>1478745</td>\n",
              "      <td>ENSG00000160072</td>\n",
              "      <td>ENST00000673477</td>\n",
              "      <td>ENSP00000500094</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>10</th>\n",
              "      <td>1</td>\n",
              "      <td>1479048</td>\n",
              "      <td>1479108</td>\n",
              "      <td>ENSG00000160072</td>\n",
              "      <td>ENST00000673477</td>\n",
              "      <td>ENSP00000500094</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>12</th>\n",
              "      <td>1</td>\n",
              "      <td>1480866</td>\n",
              "      <td>1480936</td>\n",
              "      <td>ENSG00000160072</td>\n",
              "      <td>ENST00000673477</td>\n",
              "      <td>ENSP00000500094</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>74443</th>\n",
              "      <td>1</td>\n",
              "      <td>147992501</td>\n",
              "      <td>147992567</td>\n",
              "      <td>ENSG00000188092</td>\n",
              "      <td>ENST00000314163</td>\n",
              "      <td>ENSP00000358233</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>74445</th>\n",
              "      <td>1</td>\n",
              "      <td>147992710</td>\n",
              "      <td>147992914</td>\n",
              "      <td>ENSG00000188092</td>\n",
              "      <td>ENST00000314163</td>\n",
              "      <td>ENSP00000358233</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>74451</th>\n",
              "      <td>1</td>\n",
              "      <td>147928536</td>\n",
              "      <td>147928578</td>\n",
              "      <td>ENSG00000188092</td>\n",
              "      <td>ENST00000468618</td>\n",
              "      <td>ENSP00000475253</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>74454</th>\n",
              "      <td>1</td>\n",
              "      <td>147938713</td>\n",
              "      <td>147938817</td>\n",
              "      <td>ENSG00000188092</td>\n",
              "      <td>ENST00000468618</td>\n",
              "      <td>ENSP00000475253</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>74456</th>\n",
              "      <td>1</td>\n",
              "      <td>147943437</td>\n",
              "      <td>147943544</td>\n",
              "      <td>ENSG00000188092</td>\n",
              "      <td>ENST00000468618</td>\n",
              "      <td>ENSP00000475253</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>18970 rows × 6 columns</p>\n",
              "</div>\n",
              "      <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-d8f2e8d1-0936-4b98-a844-888fdebb7ef4')\"\n",
              "              title=\"Convert this dataframe to an interactive table.\"\n",
              "              style=\"display:none;\">\n",
              "        \n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "       width=\"24px\">\n",
              "    <path d=\"M0 0h24v24H0V0z\" fill=\"none\"/>\n",
              "    <path d=\"M18.56 5.44l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94zm-11 1L8.5 8.5l.94-2.06 2.06-.94-2.06-.94L8.5 2.5l-.94 2.06-2.06.94zm10 10l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94z\"/><path d=\"M17.41 7.96l-1.37-1.37c-.4-.4-.92-.59-1.43-.59-.52 0-1.04.2-1.43.59L10.3 9.45l-7.72 7.72c-.78.78-.78 2.05 0 2.83L4 21.41c.39.39.9.59 1.41.59.51 0 1.02-.2 1.41-.59l7.78-7.78 2.81-2.81c.8-.78.8-2.07 0-2.86zM5.41 20L4 18.59l7.72-7.72 1.47 1.35L5.41 20z\"/>\n",
              "  </svg>\n",
              "      </button>\n",
              "      \n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      flex-wrap:wrap;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "      <script>\n",
              "        const buttonEl =\n",
              "          document.querySelector('#df-d8f2e8d1-0936-4b98-a844-888fdebb7ef4 button.colab-df-convert');\n",
              "        buttonEl.style.display =\n",
              "          google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "        async function convertToInteractive(key) {\n",
              "          const element = document.querySelector('#df-d8f2e8d1-0936-4b98-a844-888fdebb7ef4');\n",
              "          const dataTable =\n",
              "            await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                     [key], {});\n",
              "          if (!dataTable) return;\n",
              "\n",
              "          const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "            '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "            + ' to learn more about interactive tables.';\n",
              "          element.innerHTML = '';\n",
              "          dataTable['output_type'] = 'display_data';\n",
              "          await google.colab.output.renderOutput(dataTable, element);\n",
              "          const docLink = document.createElement('div');\n",
              "          docLink.innerHTML = docLinkHtml;\n",
              "          element.appendChild(docLink);\n",
              "        }\n",
              "      </script>\n",
              "    </div>\n",
              "  </div>\n",
              "  "
            ],
            "text/plain": [
              "      Chromosome      Start        End          gene_id    transcript_id  \\\n",
              "3              1    1471884    1472089  ENSG00000160072  ENST00000673477   \n",
              "6              1    1477273    1477350  ENSG00000160072  ENST00000673477   \n",
              "8              1    1478643    1478745  ENSG00000160072  ENST00000673477   \n",
              "10             1    1479048    1479108  ENSG00000160072  ENST00000673477   \n",
              "12             1    1480866    1480936  ENSG00000160072  ENST00000673477   \n",
              "...          ...        ...        ...              ...              ...   \n",
              "74443          1  147992501  147992567  ENSG00000188092  ENST00000314163   \n",
              "74445          1  147992710  147992914  ENSG00000188092  ENST00000314163   \n",
              "74451          1  147928536  147928578  ENSG00000188092  ENST00000468618   \n",
              "74454          1  147938713  147938817  ENSG00000188092  ENST00000468618   \n",
              "74456          1  147943437  147943544  ENSG00000188092  ENST00000468618   \n",
              "\n",
              "            protein_id  \n",
              "3      ENSP00000500094  \n",
              "6      ENSP00000500094  \n",
              "8      ENSP00000500094  \n",
              "10     ENSP00000500094  \n",
              "12     ENSP00000500094  \n",
              "...                ...  \n",
              "74443  ENSP00000358233  \n",
              "74445  ENSP00000358233  \n",
              "74451  ENSP00000475253  \n",
              "74454  ENSP00000475253  \n",
              "74456  ENSP00000475253  \n",
              "\n",
              "[18970 rows x 6 columns]"
            ]
          },
          "execution_count": 8,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "import pyranges as pr\n",
        "import pandas as pd\n",
        "# as PyRanges-object\n",
        "gr = pr.read_gtf(\"/content/Homo_sapiens.GRCh38.109.gtf\", as_df = True)\n",
        "s = (gr.Feature == \"CDS\") \n",
        "\n",
        "cds = gr[s]\n",
        "cds2 = cds[[\"Chromosome\", \"Start\", \"End\" , \"gene_id\" , \"transcript_id\" , \"protein_id\"]]\n",
        "\n",
        "cds2\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "background_save": true
        },
        "id": "V2VI3Sh4Jmi_",
        "outputId": "dec0eefa-2763-4034-9830-db66f75a2382"
      },
      "outputs": [
        {
          "data": {
            "text/html": [
              "\n",
              "  <div id=\"df-8ef39725-4689-49ff-9729-c0b3aa6d4524\">\n",
              "    <div class=\"colab-df-container\">\n",
              "      <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Chromosome</th>\n",
              "      <th>gene_id</th>\n",
              "      <th>transcript_id</th>\n",
              "      <th>protein_id</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>\n",
              "      <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-8ef39725-4689-49ff-9729-c0b3aa6d4524')\"\n",
              "              title=\"Convert this dataframe to an interactive table.\"\n",
              "              style=\"display:none;\">\n",
              "        \n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "       width=\"24px\">\n",
              "    <path d=\"M0 0h24v24H0V0z\" fill=\"none\"/>\n",
              "    <path d=\"M18.56 5.44l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94zm-11 1L8.5 8.5l.94-2.06 2.06-.94-2.06-.94L8.5 2.5l-.94 2.06-2.06.94zm10 10l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94z\"/><path d=\"M17.41 7.96l-1.37-1.37c-.4-.4-.92-.59-1.43-.59-.52 0-1.04.2-1.43.59L10.3 9.45l-7.72 7.72c-.78.78-.78 2.05 0 2.83L4 21.41c.39.39.9.59 1.41.59.51 0 1.02-.2 1.41-.59l7.78-7.78 2.81-2.81c.8-.78.8-2.07 0-2.86zM5.41 20L4 18.59l7.72-7.72 1.47 1.35L5.41 20z\"/>\n",
              "  </svg>\n",
              "      </button>\n",
              "      \n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      flex-wrap:wrap;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "      <script>\n",
              "        const buttonEl =\n",
              "          document.querySelector('#df-8ef39725-4689-49ff-9729-c0b3aa6d4524 button.colab-df-convert');\n",
              "        buttonEl.style.display =\n",
              "          google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "        async function convertToInteractive(key) {\n",
              "          const element = document.querySelector('#df-8ef39725-4689-49ff-9729-c0b3aa6d4524');\n",
              "          const dataTable =\n",
              "            await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                     [key], {});\n",
              "          if (!dataTable) return;\n",
              "\n",
              "          const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "            '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "            + ' to learn more about interactive tables.';\n",
              "          element.innerHTML = '';\n",
              "          dataTable['output_type'] = 'display_data';\n",
              "          await google.colab.output.renderOutput(dataTable, element);\n",
              "          const docLink = document.createElement('div');\n",
              "          docLink.innerHTML = docLinkHtml;\n",
              "          element.appendChild(docLink);\n",
              "        }\n",
              "      </script>\n",
              "    </div>\n",
              "  </div>\n",
              "  "
            ],
            "text/plain": [
              "Empty DataFrame\n",
              "Columns: [Chromosome, gene_id, transcript_id, protein_id]\n",
              "Index: []"
            ]
          },
          "execution_count": 11,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "\n",
        "newDf = pd.DataFrame(columns=[\"Chromosome\", \"gene_id\" , \"transcript_id\" , \"protein_id\"])\n",
        "\n",
        "copyOfr1 = df.copy()\n",
        "\n",
        "for i1, m in cds2.iterrows():\n",
        "  for i2, w in df.iterrows():\n",
        "    if (w['CHROM'] == m['Chromosome']) & (w['GeneID'] == m['gene_id']) & (w['TranscriptID'] == m['transcript_id']):\n",
        "      valueDict = {'Chromosome': w['CHROM'], 'gene_id': w['GeneID'], 'transcript_id': w['TranscriptID'], 'protein_id': m['protein_id']}\n",
        "      print(valueDict)\n",
        "      newDf = newDf.append(valueDict, ignore_index=True)\n",
        "      copyOfr1.drop(i2, inplace=True)\n",
        "      break\n",
        "\n",
        "newDf\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "7U1XYwg9JmeU"
      },
      "outputs": [],
      "source": []
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "iDUag6m4Jmbg"
      },
      "outputs": [],
      "source": []
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "SxtlZMD8JmRH"
      },
      "outputs": [],
      "source": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "TRIyA0ZsJM6f"
      },
      "source": []
    }
  ],
  "metadata": {
    "accelerator": "GPU",
    "colab": {
      "provenance": []
    },
    "gpuClass": "standard",
    "kernelspec": {
      "display_name": "Python 3",
      "name": "python3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 0
}